baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/IllegalSafeLoggingArgument.java

```diff
@@ -427,4 +432,86 @@ public Description matchClass(ClassTree tree, VisitorState state) {
     }
+
+    @Override
+    public Description matchLambdaExpression(LambdaExpressionTree tree, VisitorState state) {
+        TreePath path = state.getPath();
+        while (path != null && !(path.getLeaf() instanceof VariableTree)) {
+            path = path.getParentPath();
+        }
+        if (path == null) {
+            // The lambda isn't directly assigned to a variable, so we don't know how to analyze it (yet)
+            // TODO(aldexis): handle other ways that a lambda can be used, e.g. returned from a method
+            return Description.NO_MATCH;
+        }
+        VariableTree variable = (VariableTree) path.getLeaf();
+        Type variableType = ASTHelpers.getType(variable);
+
+        Type returnType = getFunctionalInterfaceReturnType(variableType, state);
+        if (returnType == null) {
+            return Description.NO_MATCH;
+        }
+        Safety requiredReturnSafety = SafetyAnnotations.getSafety(returnType, state);
+
+        if (requiredReturnSafety.allowsAll()) {
+            // Short-circuit if the return type allows all values
+            return Description.NO_MATCH;
+        }
+
+        Safety resultSafety = getLambdaReturnSafety(tree, state);
+
+        if (requiredReturnSafety.allowsValueWith(resultSafety)) {
+            return Description.NO_MATCH;
+        }
+
+        return buildDescription(tree)
+                .setMessage(String.format(
+                        "Dangerous return type in lambda function: expected return type is '%s' "
+                                + "but the lambda is returning '%s'.",
+                        requiredReturnSafety, resultSafety))
+                .build();
+    }
+
+    private Type getFunctionalInterfaceReturnType(Type variableType, VisitorState state) {
+        if (variableType == null) {
+            return null;
+        }
+
+        Iterable<Symbol> methods =
+                ASTHelpers.scope(variableType.tsym.members()).getSymbols(s -> s instanceof MethodSymbol);
+        List<MethodSymbol> abstractMethods = StreamSupport.stream(methods.spliterator(), false)
+                .map(MethodSymbol.class::cast)
+                .filter(m -> m.getModifiers().contains(Modifier.ABSTRACT))
+                .toList();
+        if (abstractMethods.size() != 1) {
+            // Functional interfaces should have exactly one abstract method
+            // Return nothing when we encounter something that isn't one
+            return null;
+        }
+        MethodSymbol method = Iterables.getOnlyElement(abstractMethods);
+        Type returnType = method.getReturnType();
+
+        com.sun.tools.javac.util.List<Type> classTypeArguments =
+                variableType.tsym.asType().getTypeArguments();
+
+        for (int i = 0; i < classTypeArguments.size(); i++) {
+            // TODO: check whether this is the correct way to compare the type arguments
+            if (ASTHelpers.isSameType(classTypeArguments.get(i), returnType, state)) {
```

Comment:
Not too sure about this check - I used .equals when using the expression evaluation in IntelliJ, but compilation doesn't seem happy with it, and I found this method which seemed to do what I want